core/emulate.c

```diff
@@ -571,36 +567,40 @@ static void decode_op_modrm_rm(em_context_t *ctxt,
         // Displacement
         if (ctxt->modrm.mod == 1) {
             op->mem.ea += insn_fetch_s8(ctxt);
-        }
-        if (ctxt->modrm.mod == 2) {
+        } else if (ctxt->modrm.mod == 2) {
             op->mem.ea += insn_fetch_u16(ctxt);
         }
-    }
-    if (ctxt->address_size == 4 || ctxt->address_size == 8) {
+    } else if (ctxt->address_size == 4 || ctxt->address_size == 8) {
         /* Intel SDM Vol. 2A:
          * Table 2-2. 32-Bit Addressing Forms with the ModR/M Byte */
         if (ctxt->modrm.rm == 4) {
+            uint32_t reg_base;
+            uint32_t reg_index;
+            uint8_t scale;
+
             /* Intel SDM Vol. 2A:
              * Table 2-3. 32-Bit Addressing Forms with the SIB Byte */
             ctxt->sib.value = insn_fetch_u8(ctxt);
             reg_base  = ctxt->sib.base  | (ctxt->rex.b << 3);
             reg_index = ctxt->sib.index | (ctxt->rex.x << 3);
             scale = 1 << ctxt->sib.scale;
-            op->mem.ea += READ_GPR(reg_base, ctxt->address_size);
             op->mem.ea += READ_GPR(reg_index, ctxt->address_size) * scale;
+            if ((reg_base & 7) == 5 && ctxt->modrm.mod == 0) {
+                op->mem.ea += insn_fetch_u32(ctxt);
+            } else {
+                op->mem.ea += READ_GPR(reg_base, ctxt->address_size);
+            }
         } else if (ctxt->modrm.mod == 0 && ctxt->modrm.rm == 5) {
-            op->mem.ea += insn_fetch_s32(ctxt);
+            op->mem.ea += insn_fetch_u32(ctxt);
```

Comment:
Hmm, I'd use signed int for all displacements, even if the effective address is derived from nothing but a displacement. According to Intel SDM Vol. 2A, 2.2.1.3:

> Addressing in 64-bit mode uses existing 32-bit ModR/M and SIB encodings. The ModR/M and SIB displacement
sizes do not change. They remain 8 bits or 32 bits and are **sign-extended** to 64 bits.

Reply:
Ironically, the reason I reverted it back to `u32` was because it fixed the `Booting from Floppy...` issue (as well as the GRUB menu not responding to key presses). But you are completely right, the Intel SDM is clear about this being `s32`.

Probably these are just issues canceling each other out...